src/fs_event.c

```diff
@@ -0,0 +1,103 @@
+#include "entry.h"
+
+static int fs_event_method_get(void *p, Janet key, Janet *out);
+static const JanetAbstractType fs_event_type = {
+    "uv/fs-event",
+    NULL,
+    juv_handle_mark,
+    fs_event_method_get,
+    JANET_ATEND_GET
+};
+
+static Janet cfun_fs_event_new(int32_t argc, Janet *argv) {
+    janet_fixarity(argc, 1);
+    JanetFiber *fiber = janet_getfiber(argv, 0);
+    uv_fs_event_t* handle = (uv_fs_event_t*) juv_makehandle(&fs_event_type, sizeof(*handle));
+    int ret = uv_fs_event_init(uv_default_loop(), handle);
+    if (ret < 0) janet_panic("could not create fs event monitor");
+    Janet val = juv_wrap_handle(handle);
+    janet_gcroot(val);
+    juv_handle_setfiber(handle, 0, fiber);
+    return val;
+}
+
+static void fs_event_cleanup(uv_fs_event_t *t) {
+    int res = uv_fs_event_stop(t);
+    janet_gcunroot(juv_wrap_handle(t));
+    juv_handle_setfiber(t, 0, NULL);
+    juv_handle_setcb(t, 1, NULL);
+    if (res < 0) juv_panic(res);
+}
+
+static void juv_fs_event_cb(uv_fs_event_t *handle, const char *filename, int events, int status) {
+    if (status) {
+        juv_panic(status);
+    } else {
+        JanetFunction *cb = juv_handle_cb(handle, 1);
+        Janet argv[] = { juv_wrap_handle(handle),
+                         janet_wrap_string(janet_cstring(filename)),
+                         janet_wrap_integer(events) };
+        Janet out = janet_wrap_nil();
+        JanetFiber *cb_f = janet_fiber(cb, 64, 3, argv);
+        JanetSignal sig;
+        if (!cb_f) {
+            out = janet_cstringv("arity mismatch");
+            sig = JANET_SIGNAL_ERROR;
+        } else {
+            JanetFiber *handle_f = juv_handle_fiber(handle, 0);
+            cb_f->env = handle_f->env;
+            sig = janet_continue(cb_f, janet_wrap_nil(), &out);
+        }
```

Comment:
This is a hack to ensure that the callback's fiber has an environment. If you call `janet_pcall()` without this, the fiber will have a `NULL` environment. To achieve this hack, the handle's fiber is saved to the `cbdata[0]` field. It didn't appear that anything was being put there but if that should be left blank for something else, it may cause conflicts.

Reply:
@bakpakin Do you think I should take this out? Is a fiber called with `janet_pcall()` meant to have a `NULL` environment? I mentioned this in my other comment but even with this addition, certain Janet functions which manipulate the environment (like those in Temple) can cause the fiber to crash.

Reply:
When you say crash, do you mean segfault or throw an error? Janet should handle having no fiber environment set and not segfault there.

Reply:
Sorry, I wasn't very imprecise there. I'll get a test case to make clearer what's going on and update.

Reply:
> Is a fiber called with janet_pcall() meant to have a NULL environment?

That is the current intended behavior. If you change the API to not use a callback and instead take a fiber that is resumed every time an event happens, you can avoid the problem. It should also be more efficient.

Reply:
OK. I'll have an experiment with that approach and see where I get.

Reply:
After experimenting some more, here are some thoughts:

1. In Ecstatic, I want to be able to call a function that uses `eval-string`. It seems like I can't do that with a callback function. When I change the test as follows:

   ```clj
   (defn monitor
     [parent]
     (uv/enter-loop
       (var counter 1)
       (let [callback (fn [handle path events]
                        (eval-string "(print `Hello world`)")
                        (print "File " counter " changed: " path)
                        (set counter (+ 1 counter))
                        (thread/send parent (string "Updated file: " path) 0))
             fiber    (fiber/new (fn []) :yi)
             handle   (uv/fs-event/new fiber)]
         (uv/fs-event/start handle callback "tmp" 0)
         (thread/send parent (string "Monitoring directory: " (uv/fs-event/getpath handle))))))
   ```

   I get the following error:

   ```
   error: compile error: unknown symbol print on line 1, column 1
     in compile-error [boot.janet] (tailcall) on line 2212, column 37
     in eval1 [boot.janet] on line 2146, column 16
     in run-context [boot.janet] on line 2177, column 13
     in eval-string [boot.janet] on line 2210, column 3
     in <anonymous> [test/fs-event.janet] on line 24, column 22
   uv top level signal(1): "compile error: unknown symbol print on line 1, column 1"
   ```

   If I switch to using heavy-weight threads, I get the error:

   ```
   compile error: could not add metadata to binding on line 42, column 1 while compiling test/fs-event.janet
   ```

   Is this just something that you can't do? Or is this a bug?

2. If I move the logic from the `callback` value defined in the let-block above to the `fiber` value (and make some adjustments to `fs_event_cb` to resume that fiber on an FS event rather than call the callback), I can get `eval-string` to work. However, I then have the issue of how to get data to this fiber. In `uv/timer`, a fiber works well because libuv does not have any data to pass. This isn't the case with FS events, where there is the path, the detected event and a status flag are passed ([libuv docs](http://docs.libuv.org/en/v1.x/fs_event.html#c.uv_fs_event_cb)). That information might be particularly important if, for example, the user wants to ignore changes to a particular path. The only way to pass information into the fiber when it resumes is using dynamic bindings, right? I'm not sure that is the best approach but if I wanted to do that here, how do I set a fiber's dynamic bindings from C?

3. Resuming a fiber will only work if it has not reached the end of its evaluation. Because we will typically want to respond to an arbitrary number of FS events, the only solution I could see was an endless while-loop that called `yield` at the end of each loop. That works but feels very hacky. Is that just unfortunately how it has to be?